lib/action_view/component/base.rb

```diff
@@ -189,20 +190,32 @@ def templates
             end
         end
 
-        def validate_templates
+        def valid?(raise_if_invalid)
+          if source_location.nil?
+            raise NotImplementedError.new("#{self} must implement #initialize.") if raise_if_invalid
+            return false
+          end
+
           if templates.empty?
-            raise NotImplementedError.new("Could not find a template file for #{self}.")
+            raise NotImplementedError.new("Could not find a template file for #{self}.") if raise_if_invalid
+            return false
           end
 
-          if templates.select { |template| template[:variant].nil? }.length > 1
-            raise StandardError.new("More than one template found for #{self}. There can only be one default template file per component.")
+          if templates.count { |template| template[:variant].nil? } > 1
```

Comment:
Not really crucial, but won't the invalid variants check below duplicate this check? I wonder if we could merge the two.

Reply:
We could do it that way but its a different error so I think its simpler separate.